### PR TITLE
Fix integration tests

### LINKED ISSUE #5947: Re-enable Saucelabs tests
All saucelabs jobs on all builds are failing with errors similar to this:
```
DEBUG:selenium.webdriver.remote.remote_connection:GET http://bokehplots:f7047da0-5e51-4f77-9761-cb741be552be@162.222.75.33:80/wd/hub/session/9d037f3c02894d4cb2d24fc57a1bfa33/url {"sessionId": "9d037f3c02894d4cb2d24fc57a1bfa33"}
DEBUG:selenium.webdriver.remote.remote_connection:Finished Request
DEBUG:selenium.webdriver.remote.remote_connection:GET http://bokehplots:f7047da0-5e51-4f77-9761-cb741be552be@162.222.75.33:80/wd/hub/session/9d037f3c02894d4cb2d24fc57a1bfa33/screenshot {"sessionId": "9d037f3c02894d4cb2d24fc57a1bfa33"}
DEBUG:selenium.webdriver.remote.remote_connection:Finished Request
INTERNALERROR> Traceback (most recent call last):
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/main.py", line 98, in wrap_session
INTERNALERROR>     session.exitstatus = doit(config, session) or 0
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/main.py", line 133, in _main
INTERNALERROR>     config.hook.pytest_runtestloop(session=session)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "<remote exec>", line 58, in pytest_runtestloop
INTERNALERROR>   File "<remote exec>", line 74, in run_tests
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 254, in _wrapped_call
INTERNALERROR>     return call_outcome.get_result()
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 279, in get_result
INTERNALERROR>     raise ex[1].with_traceback(ex[2])
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 265, in __init__
INTERNALERROR>     self.result = func()
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
INTERNALERROR>     res = hook_impl.function(*args)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/pytest_rerunfailures-1.0.1-py3.4.egg/pytest_rerunfailures.py", line 65, in pytest_runtest_protocol
INTERNALERROR>     reports = runtestprotocol(item, nextitem=nextitem, log=False)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/runner.py", line 79, in runtestprotocol
INTERNALERROR>     reports.append(call_and_report(item, "call", log))
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/runner.py", line 135, in call_and_report
INTERNALERROR>     report = hook.pytest_runtest_makereport(item=item, call=call)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
INTERNALERROR>     return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
INTERNALERROR>     return self._inner_hookexec(hook, methods, kwargs)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
INTERNALERROR>     _MultiCall(methods, kwargs, hook.spec_opts).execute()
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 613, in execute
INTERNALERROR>     return _wrapped_call(hook_impl.function(*args), self.execute)
INTERNALERROR>   File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 250, in _wrapped_call
INTERNALERROR>     wrap_controller.send(call_outcome)
INTERNALERROR>   File "/home/travis/build/bokeh/bokeh/tests/integration/integration_tests_plugin.py", line 139, in pytest_runtest_makereport
INTERNALERROR>     test = pytest_html.extras.image(screenshot.get_current_as_base64(), '')
INTERNALERROR>   File "/home/travis/build/bokeh/bokeh/tests/integration/screenshot.py", line 58, in get_current_as_base64
INTERNALERROR>     return self.get_screenshot_as_b64(self.current_screenshot_path)
INTERNALERROR>   File "/home/travis/build/bokeh/bokeh/tests/integration/screenshot.py", line 37, in get_screenshot_as_b64
INTERNALERROR>     with open(path, 'rb') as f:
INTERNALERROR> FileNotFoundError: [Errno 2] No such file or directory: '/home/travis/build/bokeh/bokeh/tests/integration/annotations/screenshots/current__test_label__test_label.png'
--- Logging error ---
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python3.4/logging/__init__.py", line 980, in emit
    stream.write(msg)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/capture.py", line 244, in write
    self.buffer.write(obj)
ValueError: I/O operation on closed file
Call stack:
  File "<string>", line 1, in <module>
  File "<string>", line 7, in <module>
  File "/home/travis/miniconda/lib/python3.4/site-packages/execnet/gateway_base.py", line 1469, in serve
    SlaveGateway(io=io, id=id, _startcount=2).serve()
  File "/home/travis/miniconda/lib/python3.4/site-packages/execnet/gateway_base.py", line 1014, in serve
    self._execpool.integrate_as_primary_thread()
  File "/home/travis/miniconda/lib/python3.4/site-packages/execnet/gateway_base.py", line 246, in integrate_as_primary_thread
    self._perform_spawn(reply)
  File "/home/travis/miniconda/lib/python3.4/site-packages/execnet/gateway_base.py", line 264, in _perform_spawn
    reply.run()
  File "/home/travis/miniconda/lib/python3.4/site-packages/execnet/gateway_base.py", line 200, in run
    self._result = func(*args, **kwargs)
  File "/home/travis/miniconda/lib/python3.4/site-packages/execnet/gateway_base.py", line 1039, in executetask
    do_exec(co, loc) # noqa
  File "<string>", line 1, in do_exec
  File "<remote exec>", line 154, in <module>
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/main.py", line 127, in pytest_cmdline_main
    return wrap_session(config, _main)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/main.py", line 122, in wrap_session
    exitstatus=session.exitstatus)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 745, in __call__
    return self._hookexec(self, self._nonwrappers + self._wrappers, kwargs)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 339, in _hookexec
    return self._inner_hookexec(hook, methods, kwargs)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 334, in <lambda>
    _MultiCall(methods, kwargs, hook.spec_opts).execute()
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "<remote exec>", line 38, in pytest_sessionfinish
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/vendored_packages/pluggy.py", line 614, in execute
    res = hook_impl.function(*args)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/runner.py", line 56, in pytest_sessionfinish
    session._setupstate.teardown_all()
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/runner.py", line 425, in teardown_all
    self._pop_and_teardown()
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/runner.py", line 398, in _pop_and_teardown
    self._teardown_with_finalization(colitem)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/runner.py", line 416, in _teardown_with_finalization
    self._callfinalizers(colitem)
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/runner.py", line 406, in _callfinalizers
    fin()
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/fixtures.py", line 755, in finish
    func()
  File "/home/travis/miniconda/lib/python3.4/site-packages/_pytest/fixtures.py", line 713, in teardown
    next(it)
  File "/home/travis/miniconda/lib/python3.4/site-packages/pytest_selenium-1.6.0-py3.6.egg/pytest_selenium/pytest_selenium.py", line 120, in driver
    driver.quit()
  File "/home/travis/miniconda/lib/python3.4/site-packages/selenium/webdriver/remote/webdriver.py", line 518, in quit
    self.execute(Command.QUIT)
  File "/home/travis/miniconda/lib/python3.4/site-packages/selenium/webdriver/remote/webdriver.py", line 234, in execute
    response = self.command_executor.execute(driver_command, params)
  File "/home/travis/miniconda/lib/python3.4/site-packages/selenium/webdriver/remote/remote_connection.py", line 401, in execute
    return self._request(command_info[0], url, body=data)
  File "/home/travis/miniconda/lib/python3.4/site-packages/selenium/webdriver/remote/remote_connection.py", line 415, in _request
    LOGGER.debug('%s %s %s' % (method, url, body))
Message: 'DELETE http://bokehplots:f7047da0-5e51-4f77-9761-cb741be552be@162.222.75.33:80/wd/hub/session/9d037f3c02894d4cb2d24fc57a1bfa33 {"sessionId": "9d037f3c02894d4cb2d24fc57a1bfa33"}'
Arguments: ()
```

### PR DESCRIPTION
**Pull request by canavandl**
- [x] issues: fixes #5947 
- [x] tests added / passed

### CODE REVIEW

File: bokeh/tests/test_io.py [hunk 1]
```diff
@@ -152,7 +152,36 @@ def test_missing_title(self, mock_warn):
...
+    @patch('bokeh.io.standalone_html_page_for_models')
+    def test_obj_arg_is_document(self, mock_standalone_html_page, mock_io_open):
+        obj = Document()
+        filename, resources, title = io._get_save_args(io._state, "filename", "resources", "title")
+        io._save_helper(obj, filename, resources, title, True)
+
+        self._check_func_called(mock_standalone_html_page,
+                                (obj, resources, title),
+                                {})
+        self._check_func_called(mock_io_open,
```
Comment (canavandl) [inline bokeh/tests/test_io.py]: This part of the test is "not great". It'd be better to have `io.open` generate a tempfile then check the content after writing to it, but it couldn't get it to work.

File: bokeh/io.py [hunk 2]
```diff
@@ -442,10 +442,10 @@ def _get_save_args(state, filename, resources, title):
 def _save_helper(obj, filename, resources, title, validate):
     remove_after = False
     if isinstance(obj, LayoutDOM):
-        doc = obj.document
-        if doc is None:
-            doc = Document().add_root(obj)
```
Comment (canavandl) [inline bokeh/io.py]: The problem was here, in that `Document.add_root` doesn't return a value. So trying the call `validate` on the `None` value was raising an exception and killing the integration tests.

**Comment by mattpap** [comment]
@canavandl, can you also improve integration tests, so that we get an actual exception and not some internal errors, which give no indication of really happened. Because next time I see those internal errors, I will just do `git rm` on this.

**Comment by canavandl** [comment]
I think I've figured out a route to do that (bubble the test exceptions up to the user report). I'll re-mark this PR as WIP and try to implement.

**Comment by mattpap** [comment]
@canavandl, OK, thanks.

File: tests/integration/integration_tests_plugin.py [hunk 3]
```diff
@@ -106,37 +105,23 @@ def screenshot(request):
...
-        return
-
-    # Don't add screenshots if we can't create a screenshot
-    try:
-        screenshot = Screenshot(item=item)
-    except AssertionError:
-        return
-
     report = outcome.get_result()
-    xfail = hasattr(report, 'wasxfail')
```
Comment (canavandl) [inline tests/integration/integration_tests_plugin.py]: is this currently no-op? I removed it but could add the logic back.

**Comment by canavandl** [comment]
I'm currently unable to locally reproduce the test failures (integration tests related to panning in the x-dimension). ping @bryevdv to see if you can diagnose.

**Comment by canavandl** [comment]
Overall, this causes exceptions to bubble up to the user. The html report generation continues to be the same.

New pytest log output for tests failing due to non-screenshot-related exceptions:

screen shot 2017-03-08 at 4 39 14 pm


New pytest log output for tests failing due to screenshot mismatch (you'll see on of the arrows are commented out, causing the mismatch):



screen shot 2017-03-08 at 4 39 33 pm

**Comment by mattpap** [comment]
> I'm currently unable to locally reproduce the test failures (integration tests related to panning in the x-dimension).

This is likely related to PR #5349. I will be investigating, though I'm never sure how to do this with integration tests. I would prefer to have those written in JS (plotting + interactions), so that I could run this directly in a browser and observe results.

File: tests/integration/interaction/test_range_min_max.py [hunk 4]
```diff
@@ -75,75 +56,59 @@ def pan_plot(selenium, pan_x=None, pan_y=None):
...
-    plot = make_pan_plot_with_callback(xr=Range1d(0, 3, bounds=(x_range_min, None)))
+    plot = make_plot(xr=Range1d(0, 3, bounds=(x_range_min, None)))
     save(plot)
     selenium.get(output_file_url)
     assert has_no_console_errors(selenium)
 
     # Pan plot and test for new range value
-    pan_plot(selenium, pan_x=200, pan_y=0)
-    new_range_start = float(selenium.execute_script("""alert(window.get_x_range_start())"""))
+    pan_plot(selenium, pan_x=150, pan_y=0)
```
Comment (mattpap) [inline tests/integration/interaction/test_range_min_max.py]: Why change pan range? Does if fail otherwise? This change doesn't seem related to other changes here, so it does in fact fail, maybe there is something more going on underneath.

Comment (canavandl) [inline tests/integration/interaction/test_range_min_max.py]: The tests fail without the change. One issue was that the pan end was outside of the plot frame, so that the pan event didn't happen. 

I'm not 100% sure why this wasn't a problem earlier.

Comment (bryevdv) [inline tests/integration/interaction/test_range_min_max.py]: I'm ok with this change with this explanation and @canavandl reports on local changes. any objection to merging @mattpap ?

**Comment by canavandl** [comment]
To reiterate, it's now possible to run the non-screenshot integration tests locally and use the --pdb flag as expected (which was helpful in debugging this).

It's also possible to generate the screenshot report locally via the SauceLab Connect proxy, setting the --driver=SauceLabs and --html=report.html (or any html filename and location)

File: tests/integration/interaction/test_range_min_max.py (hunk 4 above)
Comment (mattpap) [inline tests/integration/interaction/test_range_min_max.py]: Sounds reasonable, though it looks like there may be some effect of PR #5349. As long as panning works we are fine, but I will have closer look at what happens at the boundaries.

**Comment by mattpap** [comment]
LGTM

File: tests/integration/interaction/test_range_min_max.py (hunk 4 above)
Comment (bryevdv) [inline tests/integration/interaction/test_range_min_max.py]: Ok, merging now then.
